### PR TITLE
add failing skew line distanceSqToSegment test cases

### PR DESCRIPTION
**Pull request by froh**
- as per discussion in #5877
- cases from http://www.geometrictools.com/Documentation/DistanceLine3Line3.pdf

### CODE REVIEW

File: test/unit/math/Ray.js [hunk 1]
```diff
@@ -274,8 +279,38 @@ test( "distanceSqToSegment", function() {
...
+
+	// Example 1 An example where the minimum point is interior but some
+	// numerical rounding errors lead to an inaccurate result.
+
+	var P0 = new THREE.Vector3(-1.0264718499965966,	  9.6163341007195407e-007, 0.0);
+	var P1 = new THREE.Vector3( 0.91950808032415809, -1.0094441192690283e-006, 0.0);
+	var Q0 = new THREE.Vector3(-1.0629447383806110,	  9.2709540082141753e-007, 0.0);
+	var Q1 = new THREE.Vector3( 1.0811583868227901,	 -1.0670017179567367e-006, 0.0);
+
+	rayP = new THREE.Ray( P0, P1);
```
Comment (WestLangley) [inline test/unit/math/Ray.js]: The second argument in the `Ray` constructor is a direction vector, and must have unit length (be normalized). Where is `rayP` used?
This example appears to be about the distance between two line segments, which is a different problem.

File: test/unit/math/Ray.js [hunk 2]
```diff
@@ -274,8 +279,38 @@ test( "distanceSqToSegment", function() {
 	ok( ptOnSegment.distanceTo( one3 ) < 0.0001, "Passed!" );
 	ok( ptOnLine.distanceTo( one3 ) < 0.0001, "Passed!" );
 	ok( distSqr < 0.0001, "Passed!" );
-});
+	// test cases from
+	// http://www.geometrictools.com/Documentation/DistanceLine3Line3.pdf
```
Comment (WestLangley) [inline test/unit/math/Ray.js]: This paper is about the distance between 2 line segments, no?

**Comment by froh** [comment]
duh. okay.  it's (point, direction) now.  still failing.

File: test/unit/math/Ray.js [hunk 3]
```diff
@@ -274,8 +279,43 @@ test( "distanceSqToSegment", function() {
...
+	// numerical rounding errors lead to an inaccurate result.
+
+	var P0 = new THREE.Vector3(-1.0264718499965966,	  9.6163341007195407e-007, 0.0);
+	var P1 = new THREE.Vector3( 0.91950808032415809, -1.0094441192690283e-006, 0.0);
+	var dirP = new THREE.Vector3().subVectors(P1, P0);
+
+	var Q0 = new THREE.Vector3(-1.0629447383806110,	  9.2709540082141753e-007, 0.0);
+	var Q1 = new THREE.Vector3( 1.0811583868227901,	 -1.0670017179567367e-006, 0.0);
+
+	rayP = new THREE.Ray( P0, dirP);
```
Comment (WestLangley) [inline test/unit/math/Ray.js]: Where is `rayP` used?

Comment (froh) [inline test/unit/math/Ray.js]: indeed...

**Comment by froh** [comment]
the tests now both use the direction and they both now also normalize the direction vector.

the first case only exhibits the deviation if I go for a smaller threshold.  I 've increase the threshold for all cases (from 0.0001 to 0.000001), and now the two new cases both fail.

The [upstream text I pulled the examples from](http://www.geometrictools.com/Documentation/DistanceLine3Line3.pdf)
 says the first two do intersect, the second ones come much much closer than the 0.031015918671529974 when I run the tests here.

**Comment by WestLangley** [comment]
1. The paper is about testing pairs of line segments, not a ray and a segment.

2. Each pair of line segments are practically on top of one another, and nearly parallel. Their y-coordinates differ by 9e-007.

3. In your first example, the dot product between the two vectors defined by the line segments is -0.9999999999999966. In the second example, the dot product is -0.9999999999999998.

These are extremely contrived examples and are pushing the limits of the machine.

In any event, try this. Replace

    if ( det > 0 ) {

        // The ray and segment are not parallel.

with

    if ( det > 0.0000000001 ) { // or whatever

        // The ray and segment are not parallel.

**Comment by froh** [comment]
> replace
> `if ( det > 0 ) {`
> with
>  `if ( det > 0.0000000001 ) { // or whatever`

fun idea :) I've just tried that, it doesn't work out, though.

> The paper is about testing pairs of line segments, not a ray and a segment.

This is pretty similar, the sole difference is the treatment of the corner cases.

> Each pair of line segments are practically on top of one another, and nearly parallel. Their y-coordinates differ by 9e-007.

Yes, this is about near parallel lines.  Alas, the error occurs no matter how close they are, as long as they are near parallel.

> These are extremely contrived examples and are pushing the limits of the machine.

Let me try this:  I'll toy around a bit.  I check if I can come up with an alternative implementation, I look at the performance and we take it from there?

**Comment by froh** [comment]
> fun idea :) I've just tried that, it doesn't work out, though.
sorry, I did *not* try it.  I first need to figure how to get my change from src/Ray.js to the bundle that is used in the test...

But I stand with my offer to do some experimentation.

**Comment by froh** [comment]
> I first need to figure how to get my change from src/Ray.js to the bundle that is used in the test...

Done.  This is the lowest threshold that silences both cases:
`  if ( det > 1e-14 ) {`

Only  `1e-15` makes the first one fail, and from `1e-16` both fail.

I'd be happy to do just that (I also get the current math better) and focus on the RayRay stuff.

Should I just add a, say, **`1e-12`** threshold to my pull request?

**Comment by WestLangley** [comment]
The hack I suggested forces the segments to be declared parallel when they are not. The change you subsequently implemented is arbitrary, and has consequences that are not clearly understood.

As I said, your test cases are extreme. ( Do the ray and line segment intersect in the second test case, as you assume? Where do they intersect? If not, what is their true distance? )

Sorry, I recommend that this PR be closed.

If you want to suggest a replacement algorithm that is based on _careful analysis_, that would be terrific, especially if you can demonstrate a deficiency based on a reasonable test case.

**Comment by froh** [comment]
understood.

**Comment by froh** [comment]
btw the dot product of two normalized vectors is indeed 1 exactly when they are parallel, so indeed `det` is (close enough to) 0 if and only if the directions are parallel.  Which means your idea stands a *carefull analyisis* criterion.

**Pull request closed without merging**